src/core/qgstextrenderer.cpp

```diff
@@ -3550,8 +3550,28 @@ void QgsTextRenderer::drawTextInternal( TextPart drawType,
 
       bool adjustForAlignment = alignment != AlignLeft && ( mode != Label || textLines.size() > 1 );
 
-      const auto constTextLines = textLines;
-      for ( const QString &line : constTextLines )
+      // apply some character replacement to draw symbols in vertical presentation
```

Comment:
Might be nice to move this to QgsStringUtils

Reply:
@nyalldawson , good idea, will do.